core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -108,18 +108,18 @@ public RoutingNodes(ClusterState clusterState, boolean readOnly) {
                             // add the counterpart shard with relocatingNodeId reflecting the source from which
                             // it's relocating from.
                             ShardRouting targetShardRouting = shard.buildTargetRelocatingShard();
-                            addInitialRecovery(targetShardRouting, routingTable);
+                            addInitialRecovery(targetShardRouting, indexShard.primary);
                             previousValue = entries.put(targetShardRouting.shardId(), targetShardRouting);
                             if (previousValue != null) {
                                 throw new IllegalArgumentException("Cannot have two different shards with same shard id on same node");
                             }
                             assignedShardsAdd(targetShardRouting);
-                        } else if (shard.active() == false) { // shards that are initializing without being relocated
```

Comment:
here, we're iterating over the shards assigned to a node (further up outside Github view: `if (shard.assignedToNode()) {`).